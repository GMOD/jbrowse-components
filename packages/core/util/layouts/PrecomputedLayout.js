import { objectFromEntries } from '../index'

export default class PrecomputedLayout {
  constructor({ rectangles = {}, totalHeight = 0 }) {
    this.rectangles = new Map(Object.entries(rectangles))
    // rectangles is of the form "featureId": [leftPx, topPx, rightPx, bottomPx]
    this.totalHeight = totalHeight
  }

  addRect(id) {
    if (!this.rectangles.has(id)) {
      throw new Error(`id ${id} not found in precomputed feature layout`)
    }
    // left, top, right, bottom
    return this.rectangles.get(id)[1]
  }

  /**
   * returns a Map of feature id -> rectangle
   */
  getRectangles() {
    return this.rectangles
  }

  getTotalHeight() {
    return this.totalHeight
  }

  toJSON() {
    return {
      rectangles: objectFromEntries(this.rectangles),
      totalHeight: this.totalHeight,
    }
  }
}
