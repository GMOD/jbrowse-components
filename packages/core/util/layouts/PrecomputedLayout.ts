import { objectFromEntries } from '..'
import {
  RectTuple,
  SerializedLayout,
  BaseLayout,
  Rectangle,
} from './BaseLayout'
import RBush from 'rbush'

export interface Layout {
  minX: number
  minY: number
  maxX: number
  maxY: number
  name: string
}

export default class PrecomputedLayout<T> implements BaseLayout<T> {
  private rectangles: Map<string, RectTuple>

  private totalHeight: number

  public maxHeightReached: boolean

  private rbush: RBush<Layout>

  constructor({ rectangles, totalHeight, maxHeightReached }: SerializedLayout) {
    this.rectangles = new Map(Object.entries(rectangles))
    this.totalHeight = totalHeight
    this.maxHeightReached = maxHeightReached
    this.rbush = new RBush()
    for (const [key, layout] of Object.entries(rectangles)) {
      this.rbush.insert({
        minX: layout[0],
        minY: layout[1],
        maxX: layout[2],
        maxY: layout[3],
        name: key,
      })
    }
  }

  addRect(id: string) {
    const rect = this.rectangles.get(id)
    if (!rect) {
      throw new Error(`id ${id} not found in precomputed feature layout`)
    }
    // left, top, right, bottom
    return rect[1]
  }

  /**
   * returns a Map of `feature id -> rectangle`
   */
  getRectangles(): Map<string, RectTuple> {
    return this.rectangles
  }

  getTotalHeight(): number {
    return this.totalHeight
  }

  collides(_rect: Rectangle<T>, _top: number): boolean {
    throw new Error('Method not implemented.')
  }

  getByCoord(x: number, y: number) {
    const rect = { minX: x, minY: y, maxX: x + 1, maxY: y + 1 }
    return this.rbush.collides(rect)
      ? this.rbush.search(rect)[0].name
      : undefined
  }

  getByID(id: string) {
    return this.rectangles.get(id)
  }

  addRectToBitmap(_rect: Rectangle<T>, _data: Record<string, T>): void {
    throw new Error('Method not implemented.')
  }

  discardRange(_left: number, _right: number): void {
    throw new Error('Method not implemented.')
  }

  serializeRegion(_region: { start: number; end: number }): SerializedLayout {
    throw new Error('Method not implemented.')
  }

  toJSON(): SerializedLayout {
    return {
      rectangles: objectFromEntries(this.rectangles),
      totalHeight: this.totalHeight,
      maxHeightReached: false,
    }
  }
}
