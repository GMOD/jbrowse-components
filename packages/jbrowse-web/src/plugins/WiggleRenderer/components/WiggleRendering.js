import React, { Component } from 'react'
import { observer } from 'mobx-react'
import ReactPropTypes from 'prop-types'

import './WiggleRendering.scss'
import PrerenderedCanvas from './PrerenderedCanvas'

import { PropTypes as CommonPropTypes } from '../../../mst-types'

class WiggleRendering extends Component {
  static propTypes = {
    height: ReactPropTypes.number.isRequired,
    width: ReactPropTypes.number.isRequired,
    region: CommonPropTypes.Region.isRequired,
    bpPerPx: ReactPropTypes.number.isRequired,
    horizontallyFlipped: ReactPropTypes.bool,
    trackModel: ReactPropTypes.shape({}),
  }

  static defaultProps = {
    horizontallyFlipped: false,
    trackModel: {},
  }

  constructor(props) {
    super(props)
    this.state = {}
  }

  onMouseMove(evt) {
    const { region, features, bpPerPx, horizontallyFlipped, width } = this.props
    const { offsetX } = evt.nativeEvent
    const px = horizontallyFlipped ? width - offsetX : offsetX
    const clientBp = region.start + bpPerPx * px
    for (const feature of features.values()) {
      if (clientBp <= feature.get('end') && clientBp >= feature.get('start')) {
        this.setState({ offsetX, feature })
        return
      }
    }
  }

  onMouseLeave() {
    this.setState({ feature: undefined })
  }

  render() {
    const { width } = this.props
    const { feature, offsetX } = this.state
    return (
      <>
        <div
          onMouseMove={this.onMouseMove.bind(this)}
          onMouseLeave={this.onMouseLeave.bind(this)}
          onFocus={() => {}}
          className="WiggleRendering"
          style={{ position: 'relative' }}
        >
          <PrerenderedCanvas {...this.props} width={width} />
          {feature ? (
            <div style={{ pointerEvents: 'none' }}>
              <div className="hoverLabel" style={{ left: `${offsetX}px` }}>
                {parseFloat(feature.get('score').toPrecision(6))}
              </div>
              <div className="hoverVertical" style={{ left: `${offsetX}px` }} />
            </div>
          ) : null}
        </div>
      </>
    )
  }
}

export default observer(WiggleRendering)
