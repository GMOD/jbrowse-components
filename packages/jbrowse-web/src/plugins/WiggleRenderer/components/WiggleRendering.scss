.WiggleRendering {
  .blur {
    text-align: center;
    color: black;
  }

}
