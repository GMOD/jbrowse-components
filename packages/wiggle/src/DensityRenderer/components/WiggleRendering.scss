.WiggleRendering {
  .blur {
    text-align: center;
    color: black;
  }

  .hoverVertical {
    background: #333;
    border: none;
    width: 1px;
    height: 100%;
    top: 0px;
    cursor: default;
    position: absolute;
  }
  .hoverLabel {
    border: 1px solid black;
    background-color: white;
    top: 0px;
    position: absolute;
  }
}
