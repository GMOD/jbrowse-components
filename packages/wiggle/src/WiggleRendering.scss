.WiggleRendering {
  .blur {
    text-align: center;
    color: black;
  }

  .hoverVertical {
    background: #333;
    border: none;
    width: 1px;
    height: 100%;
    top: 0px;
    cursor: default;
    position: absolute;
  }
  .hoverLabel {
    border: 1px solid black;
    background-color: white;
    top: 0px;
    position: absolute;
  }

  table,
  th,
  td {
    font-size: 10px;
    border: 1px solid black;
    border-collapse: collapse;
  }

  #smallInfo {
    font-size: 8px;
  }
}
