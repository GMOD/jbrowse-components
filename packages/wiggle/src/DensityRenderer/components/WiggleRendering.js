import { PropTypes as CommonPropTypes } from '@gmod/jbrowse-core/mst-types'
import PrerenderedCanvas from '@gmod/jbrowse-core/components/PrerenderedCanvas'
import { observer } from 'mobx-react'
import ReactPropTypes from 'prop-types'
import React, { Component } from 'react'
import './WiggleRendering.scss'

class WiggleRendering extends Component {
  static propTypes = {
    height: ReactPropTypes.number.isRequired,
    width: ReactPropTypes.number.isRequired,
    region: CommonPropTypes.Region.isRequired,
    features: ReactPropTypes.instanceOf(Map).isRequired,
    bpPerPx: ReactPropTypes.number.isRequired,
    horizontallyFlipped: ReactPropTypes.bool,
    trackModel: ReactPropTypes.shape({
      /** id of the currently selected feature, if any */
      selectedFeatureId: ReactPropTypes.string,
    }),
  }

  static defaultProps = {
    horizontallyFlipped: false,
    trackModel: {},
  }

  constructor(props) {
    super(props)
    this.state = {}
  }

  onMouseMove(evt) {
    const { region, features, bpPerPx, horizontallyFlipped, width } = this.props
    const { offsetX } = evt.nativeEvent
    const px = horizontallyFlipped ? width - offsetX : offsetX
    const clientBp = region.start + bpPerPx * px
    for (const feature of features.values()) {
      if (clientBp <= feature.get('end') && clientBp >= feature.get('start')) {
        this.setState({ offsetX, featureUnderMouse: feature })
        return
      }
    }
  }

  onMouseLeave() {
    this.setState({ featureUnderMouse: undefined })
  }

  /**
   * @param {string} handlerName
   * @param {*} event - the actual mouse event
   * @param {bool} always - call this handler even if there is no feature
   */
  callMouseHandler(handlerName, event, always = false) {
    // eslint-disable-next-line react/destructuring-assignment
    const featureHandler = this.props[`onFeature${handlerName}`]
    // eslint-disable-next-line react/destructuring-assignment
    const canvasHandler = this.props[`on${handlerName}`]
    const { featureUnderMouse } = this.state
    if (featureHandler && (always || featureUnderMouse)) {
      featureHandler(event, featureUnderMouse)
    } else if (canvasHandler) {
      canvasHandler(event, featureUnderMouse)
    }
  }

  render() {
    const { featureUnderMouse, offsetX } = this.state
    const { height } = this.props

    const toP = s => parseFloat(s.toPrecision(6))
    const getFeatRepr = feature => {
      return feature.get('maxScore') !== undefined ? (
        <div>
          Summary
          <br />
          Max: {toP(feature.get('maxScore'))}
          <br />
          Avg: {toP(feature.get('score'))}
          <br />
          Min: {toP(feature.get('minScore'))}
        </div>
      ) : (
        toP(feature.get('score'))
      )
    }
    const getMouseoverFlag = feature => (
      <div style={{ pointerEvents: 'none' }}>
        <div className="hoverLabel" style={{ left: `${offsetX}px` }}>
          {getFeatRepr(feature)}
        </div>
        <div className="hoverVertical" style={{ left: `${offsetX}px` }} />
      </div>
    )
    return (
      <div
        onMouseMove={this.onMouseMove.bind(this)}
        onMouseLeave={this.onMouseLeave.bind(this)}
        role="presentation"
        onFocus={() => {}}
        className="WiggleRendering"
        style={{
          overflow: 'hidden',
          height,
        }}
      >
        <PrerenderedCanvas {...this.props} />
        {featureUnderMouse ? getMouseoverFlag(featureUnderMouse) : null}
      </div>
    )
  }
}

export default observer(WiggleRendering)
