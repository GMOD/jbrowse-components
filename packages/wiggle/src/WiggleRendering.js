import { PropTypes as CommonPropTypes } from '@gmod/jbrowse-core/mst-types'
import { PrerenderedCanvas } from '@gmod/jbrowse-core/ui'
import { observer } from 'mobx-react'
import ReactPropTypes from 'prop-types'
import React, { Component } from 'react'
import './WiggleRendering.scss'

const toP = s => parseFloat(s.toPrecision(6))

function Tooltip({ offsetX, feature }) {
  return (
    <>
      <div
        className="hoverLabel"
        style={{ left: `${offsetX}px`, zIndex: 10000 }}
      >
        {feature.get('maxScore') !== undefined ? (
          <div>
            Summary
            <br />
            Max: {toP(feature.get('maxScore'))}
            <br />
            Avg: {toP(feature.get('score'))}
            <br />
            Min: {toP(feature.get('minScore'))}
          </div>
        ) : (
          toP(feature.get('score'))
        )}
      </div>
      <div className="hoverVertical" style={{ left: `${offsetX}px` }} />
    </>
  )
}

Tooltip.propTypes = {
  offsetX: ReactPropTypes.number.isRequired,
  feature: ReactPropTypes.shape({ get: ReactPropTypes.func }).isRequired,
}

class WiggleRendering extends Component {
  static propTypes = {
    height: ReactPropTypes.number.isRequired,
    width: ReactPropTypes.number.isRequired,
    region: CommonPropTypes.Region.isRequired,
    features: ReactPropTypes.instanceOf(Map).isRequired,
    bpPerPx: ReactPropTypes.number.isRequired,
    horizontallyFlipped: ReactPropTypes.bool,
    trackModel: ReactPropTypes.shape({
      /** id of the currently selected feature, if any */
      selectedFeatureId: ReactPropTypes.string,
    }),
  }

  static defaultProps = {
    horizontallyFlipped: false,
    trackModel: {},
  }

  constructor(props) {
    super(props)
    this.state = {}
    this.ref = React.createRef()
  }

  onMouseMove(evt) {
    const { region, features, bpPerPx, horizontallyFlipped, width } = this.props
    const { clientX } = evt
    let offset = 0
    if (this.ref.current) {
      offset = this.ref.current.getBoundingClientRect().left
    }
    const offsetX = clientX - offset
    const px = horizontallyFlipped ? width - offsetX : offsetX
    const clientBp = region.start + bpPerPx * px

    for (const feature of features.values()) {
      if (clientBp <= feature.get('end') && clientBp >= feature.get('start')) {
        this.setState({ clientX, featureUnderMouse: feature })
        return
      }
    }
  }

  onMouseLeave() {
    this.setState({ featureUnderMouse: undefined })
  }

  /**
   * @param {string} handlerName
   * @param {*} event - the actual mouse event
   * @param {bool} always - call this handler even if there is no feature
   */
  callMouseHandler(handlerName, event, always = false) {
    // eslint-disable-next-line react/destructuring-assignment
    const featureHandler = this.props[`onFeature${handlerName}`]
    // eslint-disable-next-line react/destructuring-assignment
    const canvasHandler = this.props[`on${handlerName}`]
    const { featureUnderMouse } = this.state
    if (featureHandler && (always || featureUnderMouse)) {
      featureHandler(event, featureUnderMouse)
    } else if (canvasHandler) {
      canvasHandler(event, featureUnderMouse)
    }
  }

  render() {
    const { featureUnderMouse, clientX } = this.state
    const { height } = this.props
    let offset = 0
    if (this.ref.current) {
      offset = this.ref.current.getBoundingClientRect().left
    }

    return (
      <div
        ref={this.ref}
        onMouseMove={this.onMouseMove.bind(this)}
        onMouseLeave={this.onMouseLeave.bind(this)}
        role="presentation"
        onFocus={() => {}}
        className="WiggleRendering"
        style={{
          overflow: 'visible',
          position: 'relative',
          height,
        }}
      >
        <PrerenderedCanvas {...this.props} />
        {featureUnderMouse ? (
          <Tooltip feature={featureUnderMouse} offsetX={clientX - offset} />
        ) : null}
      </div>
    )
  }
}

export default observer(WiggleRendering)
