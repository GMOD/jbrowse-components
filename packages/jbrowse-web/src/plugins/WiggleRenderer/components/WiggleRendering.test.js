import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import WiggleRendering from './WiggleRendering'
import PrecomputedLayout from '../../../util/layouts/PrecomputedLayout'

// these tests do very little, let's try to expand them at some point
test('one', () => {
  const renderer = new ShallowRenderer()
  renderer.render(
    <WiggleRendering
      width={500}
      height={500}
      region={{ assemblyName: 'wookie', refName: 'chr1', start: 1, end: 3 }}
      layout={new PrecomputedLayout({ rectangles: {}, totalHeight: 20 })}
      bpPerPx={3}
    />,
  )
  const result = renderer.getRenderOutput()

  expect(result).toMatchSnapshot()
})
