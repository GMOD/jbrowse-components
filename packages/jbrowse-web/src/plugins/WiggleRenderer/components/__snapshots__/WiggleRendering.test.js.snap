// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`one 1`] = `
<React.Fragment>
  <div
    className="WiggleRendering"
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <PrerenderedCanvas
      bpPerPx={3}
      height={500}
      horizontallyFlipped={false}
      region={
        Object {
          "end": 3,
          "refName": "chr1",
          "start": 1,
        }
      }
      scale={2}
      trackModel={Object {}}
      width={500}
    />
  </div>
</React.Fragment>
`;
